src/libraries/System.Text.Json/docs/JsonSerializationSourceGeneration.md

```diff
@@ -0,0 +1,238 @@
+# System.Text.Json build-time generation for serializers using SourceGenerators
+
+## Motivation
+
+There are comprehensive [documents](https://github.com/dotnet/designs/pull/113) detailing the needs and benefits of generating JSON serializers at compile time. Some of these benefits are faster throughput, **improved startup time**, and **reduction in private memory usage** for serialization/deserialization. After discussing some approaches and pros/cons of some of them we decided to go ahead and implement this feature using [Roslyn Source Generators](https://github.com/dotnet/roslyn/blob/master/docs/features/source-generators.cookbook.md). This document will outline the roadmap for the initial experiment and highlight actionable items for the base prototype.
+
+## Feature Behavior
+There are 3 main points in this project: type discovery, source code generation and generated source code integration (with user applications):
+
+### Type Discovery
+
+Type discovery can be divided into two different models, an implicit model (where the user does not have to specify which types to generate code for) and an explicit model (user specifically specifies through code which types they want us to generate code for). 
+
+#### Implicit Model
+Various implicit approaches have been discussed such as source generating all partial classes or scanning for calls into the serializer using the Roslyn tree scanning. These models can be revisited in the future as the value/feasibility of the approach becomes clearer based on user feedback. It is important to note that some downsides to such model can result in missing types to source generate or source generating types when not needed due to a bug or edge cases we didn’t consider. 
+
+#### Proposed Explicit Model
+There are two scenarios within the proposed explicit model:
+
+1. Base case: Code generates a facade ```JsonClassInfo``` for the attribute target class/struct.
+
+```c#
+ // (Base Case) Codegen (De)Serialization ExampleLoginClassInfo. 
+[JsonSerializable] 
```

Comment:
Looks like a good start, but cant we also be able to pass the ``SerializerOptions`` we want the source generator to use to this attribute as well too? Like if we know the json contains a lot of things where setting specific options would work, but the default ones would not work at all.

Reply:
Yes, the generated metadata context class provides a ctor that will accept a `JsonSerializerOptions` instance which will be honored at runtime. See [this](https://github.com/layomia/JsonSourceGenPerf/blob/cf83d7814b18bd72b4847abee13106899a2abe53/SizeOpts/BlazorAfter/Pages/FetchData.razor#L40-L41) as an example.